trie/trieStorageManager.go

```diff
@@ -289,37 +281,66 @@ func (tsm *trieStorageManager) takeSnapshot(snapshotEntry *snapshotsQueueEntry,
 		}
 	}()
 
+	var db temporary.DBWriteCacher
 	if tsm.isPresentInLastSnapshotDb(snapshotEntry.rootHash) {
-		log.Trace("snapshot for rootHash already taken", "rootHash", snapshotEntry.rootHash)
-		return
+		db = tsm.GetSnapshotThatContainsHash(snapshotEntry.rootHash)
+		isDbNil := check.IfNil(db)
+		log.Trace("snapshot for rootHash already taken, using snapshot DB",
+			"rootHash", snapshotEntry.rootHash, "is DB nil", isDbNil)
 	}
 
-	log.Trace("trie snapshot started", "rootHash", snapshotEntry.rootHash, "newDB", snapshotEntry.newDb)
+	if check.IfNil(db) {
+		log.Trace("source DB is nil, setting the source DB as tsm.db")
+		db = tsm.db
+	}
+	log.Trace("trie checkpoint started", "rootHash", snapshotEntry.rootHash)
 
-	newRoot, err := newSnapshotNode(tsm.db, msh, hsh, snapshotEntry.rootHash)
+	newRoot, err := newSnapshotNode(db, msh, hsh, snapshotEntry.rootHash)
 	if err != nil {
 		log.Error("trie storage manager: newSnapshotTrie", "error", err.Error())
 		return
 	}
-	db := tsm.getSnapshotDb(snapshotEntry.newDb)
-	if check.IfNil(db) {
+	newDb := tsm.getSnapshotDb(snapshotEntry.newDb)
+	if check.IfNil(newDb) {
 		return
 	}
 
-	if snapshotEntry.entryType == snapshot {
-		err = newRoot.commitSnapshot(tsm.db, db, snapshotEntry.leavesChan, ctx)
-		if err == ErrContextClosing {
-			log.Debug("context closing while in commitSnapshot operation")
-			return
-		}
-		if err != nil {
-			log.Error("trie storage manager: commit", "error", err.Error())
+	err = newRoot.commitSnapshot(db, newDb, snapshotEntry.leavesChan, ctx)
+	if err == ErrContextClosing {
+		log.Debug("context closing while in commitSnapshot operation")
+		return
+	}
+	if err != nil {
+		log.Error("trie storage manager: commit", "error", err.Error())
+	}
+}
+
+func (tsm *trieStorageManager) takeCheckpoint(checkpointEntry *snapshotsQueueEntry, msh marshal.Marshalizer, hsh hashing.Hasher, ctx context.Context) {
```

Comment:
some unit tests for the new functions ?

Reply:
the existing unit test will test the new function as the existing functionality was already tested. Though, the code coverage is not great. Adding more unit tests...